### ISSUE TITLE
CentOS 6 build of ngx_pagespeed uses huge amounts of CPU.

### ISSUE THREAD

**Issue by sholdowa** [pagespeed/ngx_pagespeed#975]
I've built a clean version of ngx_pagespeed with nginx 1.9.1. To do this, I based it upon the standard build for CentOS so that I can switch back and forth from the repo version of nginx, whiich works perfectly. The build script ended up as:

./configure --prefix=/etc/nginx \
--sbin-path=/usr/sbin/nginx \
--conf-path=/etc/nginx/nginx.conf \
--error-log-path=/var/log/nginx/error.log \
--http-log-path=/var/log/nginx/access.log \
--pid-path=/var/run/nginx.pid \
--lock-path=/var/run/nginx.lock \
--http-client-body-temp-path=/var/cache/nginx/client_temp \
--http-proxy-temp-path=/var/cache/nginx/proxy_temp \
--http-fastcgi-temp-path=/var/cache/nginx/fastcgi_temp \
--http-uwsgi-temp-path=/var/cache/nginx/uwsgi_temp \
--http-scgi-temp-path=/var/cache/nginx/scgi_temp \
--user=nginx \
--group=nginx \
--with-ipv6 \
--with-http_ssl_module \
--with-http_spdy_module \
--with-http_realip_module \
--with-http_addition_module \
--with-http_sub_module \
--with-http_gunzip_module \
--with-http_gzip_static_module \
--with-http_random_index_module \
--with-http_secure_link_module \
--with-http_stub_status_module \
--with-mail \
--with-mail_ssl_module \
--with-file-aio \
--add-module=../ngx_pagespeed-release-1.9.32.3-beta \
'--with-cc-opt=-O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic '

This builds and runs fine *unless* I attempt to use ngx_pagespeed.

Using the minimal config

pagespeed on;

# Needs to exist and be writable by nginx.  Use tmpfs for best performance.
pagespeed FileCachePath /var/cache/pagespeed/;

# Ensure requests for pagespeed optimized resources go to the pagespeed handler
# and no extraneous headers get set.
location ~ "\.pagespeed\.([a-z]\.)?[a-z]{2}\.[^.]{10}\.[^.]+" {
  add_header "" "";
}
location ~ "^/pagespeed_static/" { }
location ~ "^/ngx_pagespeed_beacon$" { }

in a single site (  /var/cache/pagespeed/ is a 1GB tmpfs mounted as apache ), I immediately get 2 nginx processes running at 100%, and they stay there. THere's nothing extra being written to the cache dir...

$ du -sk /var/cache/pagespeed/*
964	/var/cache/pagespeed/diamondharbour.org.nz
4	/var/cache/pagespeed/prop_page
84	/var/cache/pagespeed/rname

WHat's happening???

**Comment by oschaaf** [comment]
@sholdowa Reading up on https://github.com/pagespeed/ngx_pagespeed/issues/956 might help

**Comment by sholdowa** [comment]
@oschaaf 

As I explained, FileSystemCache is configured, writeable and being populated.
SELinux is disabled

The only pagespeed related messages in the nginx error log are
2015/06/06 11:37:27 [info] 18602#0: [ngx_pagespeed 1.9.32.3-4448] No threading detected. Own threads: 1 Rewrite, 1 Expensive Rewrite.
2015/06/06 11:37:34 [info] 18608#0: [ngx_pagespeed 1.9.32.3-4448] No threading detected. Own threads: 1 Rewrite, 1 Expensive Rewrite.

Any other ideas? Does enabling pagecache on a single site enable it for all virtual hosts, leading the others to have an incorrect FileSystemCache maybe??

**Comment by sholdowa** [comment]
Hmm... I might have hit the nail on the head with the last comment...

If I add the line

pagespeed FileCachePath /var/cache/pagespeed/;

as http: scope, but configure it nowhere, I get the same 100% CPU usage for nginx. Surely that's not right???

**Comment by oschaaf** [comment]
@sholdowa There's also a case where it is possible to get into an cpu spin on centos 6, when gcc / ld  have a mismatch. See https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/mod-pagespeed-discuss/1EkahAxLg-s/dcC6jEeReVQJ

TL;DR:
```
Fixes for user:
Confirmed: compile pagespeed from source and link against that instead of using the downloadable static library, this will remove the .ctors section.
Confirmed: upgrade ld, so mixed .ctors and .init_array will be moved to .init_array by ld and the entries will be executed in expected order
downgrade gcc, so .ctor entries will be generated and not .init_array
```

**Comment by sholdowa** [comment]
That might be a possibility, but

________ running 'git -c core.deltaBaseCacheLimit=2g clone --no-checkout --progress https://github.com/pagespeed/mod_pagespeed.git /usr/local/src/mod_pagespeed/_gclient_src_uwHv6S' in '/usr/local/src/mod_pagespeed'
[0:00:00] Unknown option: -c
[0:00:00] usage: git [--version] [--exec-path[=GIT_EXEC_PATH]] [--html-path]
[0:00:00]            [-p|--paginate|--no-pager] [--no-replace-objects]
[0:00:00]            [--bare] [--git-dir=GIT_DIR] [--work-tree=GIT_WORK_TREE]
[0:00:00]            [--help] COMMAND [ARGS]

The build tools don't seem to work under CentOS 6.

**Comment by crowell** [comment]
@sholdowa my guess is that your version of git is too old.

if your version of git is < 1.8.x try installing from source. see the instructions here

https://developers.google.com/speed/pagespeed/module/build_mod_pagespeed_from_source

**Comment by crowell** [comment]
seems that centos6 does ship with git 1.7.1 (from http://mirror.centos.org/centos/6/os/x86_64/Packages/).
Installing a newer version of git should fix thiat for you.

**Comment by keesspoelstra** [comment]
One word of advice, do not install the python 2.7 on Centos 6 the way as it is described for Centos 5 in the mod_pagespeed article, this will break yum and some other system tools.
Install it by installing the scl tools:
http://wiki.centos.org/AdditionalResources/Repositories/SCL
